--- cursors.py.orig	2013-06-04 09:28:10.659558016 +1000
+++ cursors.py	2013-06-04 09:28:23.095321467 +1000
@@ -49,7 +49,7 @@
         try:
             while self.nextset():
                 pass
-        except:
+        except Exception:
             pass
 
         self.connection = None
@@ -110,7 +110,7 @@
         result = 0
         try:
             result = self._query(query)
-        except:
+        except Exception:
             exc, value, tb = exc_info()
             del tb
             self.messages.append((exc,value))
